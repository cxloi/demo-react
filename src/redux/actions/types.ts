// type - redux recommends naming it as event

export const INCREASE_COUNT = "INCREASE_COUNT";
export const DECREASE_COUNT = "DECREASE_COUNT";

export const USER = {
  GET_DETAILS_REQUEST: "USER/GET_DETAILS_REQUEST",
  GET_DETAILS_RESPONSE: "USER/GET_DETAILS_RESPONSE",
};
