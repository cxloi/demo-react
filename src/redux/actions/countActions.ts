import { DECREASE_COUNT, INCREASE_COUNT } from './types';

interface IncreaseCountAction {
  type: typeof INCREASE_COUNT;
  // payload
}

interface DecreaseCountAction {
  type: typeof DECREASE_COUNT;
}

export type CountActionTypes = IncreaseCountAction | DecreaseCountAction;

export function increaseCountAction(): IncreaseCountAction {
  return {
    type: INCREASE_COUNT,
  };
}

export function decreaseCountAction(): DecreaseCountAction {
  return {
    type: DECREASE_COUNT,
  };
}
