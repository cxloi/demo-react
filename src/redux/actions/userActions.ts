import { UserDetails } from '../../model/UserModel';
import { USER_LOGGED_IN_SUCCESS, USER_LOGGED_OUT_SUCCESS, USER_UPDATE_DETAILS } from './types';

interface UserLoggedInSuccessAction {
  type: typeof USER_LOGGED_IN_SUCCESS;
  payload: string;
}

interface UserLoggedOutSuccessAction {
  type: typeof USER_LOGGED_OUT_SUCCESS;
}

interface UserUpdateDetailsAction {
  type: typeof USER_UPDATE_DETAILS;
  payload: UserDetails;
}

export type UserActionTypes =
  | UserLoggedInSuccessAction
  | UserLoggedOutSuccessAction
  | UserUpdateDetailsAction;

export function userLoggedInSuccess(userId: string): UserActionTypes {
  return {
    type: USER_LOGGED_IN_SUCCESS,
    payload: userId,
  };
}

export function userLoggedOutSuccess(): UserActionTypes {
  return {
    type: USER_LOGGED_OUT_SUCCESS,
  };
}

export function userUpdateDetails(userDetails: UserDetails): UserActionTypes {
  return {
    type: USER_UPDATE_DETAILS,
    payload: userDetails,
  };
}
