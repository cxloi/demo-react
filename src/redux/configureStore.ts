import {
  // applyMiddleware,
  // compose,
  createStore,
} from 'redux';

// import monitorReducersEnhancer from "./enhancers/monitorReducers";
// import loggerMiddleware from "./middleware/logger";
import { rootReducer } from './reducers';

export default function configureStore() {
  // const middlewares = [loggerMiddleware, thunkMiddleware]
  // const middlewareEnhancer = applyMiddleware(...middlewares)

  // const enhancers = [middlewareEnhancer, monitorReducersEnhancer]
  // const composedEnhancers = compose(...enhancers)

  const store = createStore(
    rootReducer,
    // preloadedState,
    // composedEnhancers
  );

  return store;
}
