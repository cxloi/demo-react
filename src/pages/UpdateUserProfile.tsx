import { CustomButton } from 'demo-sb-react-components';
import React, { CSSProperties, useEffect, useReducer } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { UserDetails } from '../model/UserModel';
import { userUpdateDetails } from '../redux/actions/userActions';
import { selectIsUserLoggedIn, selectUserDetails } from '../redux/selectors';

const inputContainerStyle: CSSProperties = {
  marginTop: 5,
  marginBottom: 15,
};

const USER_FORM_ACTIONS = {
  UPDATE: 'USER_FORM_ACTIONS/UPDATE',
};

interface UserDetailsFormState extends UserDetails {}

interface UserDetailsFormAction {
  type: typeof USER_FORM_ACTIONS.UPDATE;
  payload: UserDetailsFormState;
}

const userDetailsFormReducer = (
  state: UserDetailsFormState,
  action: UserDetailsFormAction,
): UserDetailsFormState => {
  switch (action.type) {
    case USER_FORM_ACTIONS.UPDATE:
      return { ...state, ...action.payload };
    default:
      return { ...state };
  }
};

const UpdateUserProfile = () => {
  const dispatch = useDispatch();
  const isUserLoggedIn = useSelector(selectIsUserLoggedIn);
  const userDetails = useSelector(selectUserDetails);

  const [state, userDetailsFormDispatch] = useReducer(userDetailsFormReducer, userDetails);
  const { userName = '', age = 0 } = state;

  useEffect(() => {
    userDetailsFormDispatch({ type: USER_FORM_ACTIONS.UPDATE, payload: userDetails });
  }, [userDetails]);

  return (
    <div data-testid="updatePage">
      <h2>Update User Profile</h2>

      <div>
        {isUserLoggedIn ? (
          <>
            <div>User Name:</div>
            <input
              type="text"
              aria-label="input-custom-name"
              value={userName}
              onChange={(event) =>
                userDetailsFormDispatch({
                  type: USER_FORM_ACTIONS.UPDATE,
                  payload: { userName: event.target.value },
                })
              }
              style={inputContainerStyle}
            />

            <div>Age:</div>
            <input
              value={age}
              onChange={(event) =>
                userDetailsFormDispatch({
                  type: USER_FORM_ACTIONS.UPDATE,
                  payload: { age: event.target.valueAsNumber },
                })
              }
              type="number"
              style={inputContainerStyle}
            />

            <br />
            <CustomButton
              data-testid="updateButton"
              label="Submit"
              onClick={() => dispatch(userUpdateDetails(state))}
            />
          </>
        ) : (
          <div>Please log in to update user profile.</div>
        )}
      </div>
    </div>
  );
};

export default UpdateUserProfile;
