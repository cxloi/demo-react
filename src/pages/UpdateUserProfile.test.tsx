import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { userUpdateDetails } from '../redux/actions/userActions';
import UpdateUserProfile from './UpdateUserProfile';

const mockState = {
  count: 1,
  user: {
    isLoggedIn: true,
    userDetails: {
      userName: 'Dan Abramov',
      age: 38,
    },
  },
};

const mockStore = configureStore([])(mockState);

const renderComponent = () => render(<Provider store={mockStore}><UpdateUserProfile /></Provider>);

describe('UpdateUserProfile', () => {
  test('"render update form', () => {
    const { getByTestId } = renderComponent();
    const textElement = getByTestId('updatePage').textContent;
    expect(textElement).toContain('User Name:');
    expect(textElement).toContain('Age:');
    expect(textElement).toContain('Submit');
  });

  test('Update action dispatch to store correctly', () => {
    const action = userUpdateDetails({ userName: 'Michael', age: 40 });
    mockStore.dispatch(action);

    const actions = mockStore.getActions();
    expect(actions).toEqual([action]);
  });
});
