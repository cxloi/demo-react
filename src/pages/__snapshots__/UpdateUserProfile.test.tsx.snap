// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateUserProfile render snapshot 1`] = `
<div
  data-testid="updatePage"
>
  <h2>
    Update User Profile
  </h2>
  <div>
    <div>
      User Name:
    </div>
    <input
      aria-label="input-custom-name"
      onChange={[Function]}
      style={
        Object {
          "marginBottom": 15,
          "marginTop": 5,
        }
      }
      type="text"
      value="Dan Abramov"
    />
    <div>
      Age:
    </div>
    <input
      onChange={[Function]}
      style={
        Object {
          "marginBottom": 15,
          "marginTop": 5,
        }
      }
      type="number"
      value={38}
    />
    <br />
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-container-3 storybook-button storybook-button--medium MuiButton-containedPrimary"
      data-testid="updateButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": undefined,
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Submit
      </span>
    </button>
  </div>
</div>
`;
